### ISSUE TITLE
useHover sometimes never called on leave

### ISSUE THREAD

**Issue by jwalton** [pmndrs/use-gesture#322]
**Describe the bug**

Sometimes the hover gesture never gets called with `isHovering: false` when the mouse leaves the element in question.

**Sandbox or Video**

I do not have a minimal reproduction (yet).  :(

Here's a video that shows the problem:

![no-hover-leave](https://user-images.githubusercontent.com/1771003/122958012-79d86000-d350-11eb-8e23-51adf116315e.gif)

What you're looking at here is a bunch of "events" that happened on a video, but there's too many events to show, so when you mouse over the events they climb up to cover over the video controls, and then when you mouse out they're supposed to collapse back down.  As you can see here, though, sometimes they do not.

Since I also want to know what time the user is currently hovering over on that timeline, this is all based on this hook to give me the relative `x,y` position of the mouse:

```ts
export function useRelativeMousePosition(
  ref: Element | React.RefObject<Element>,
  options?: {
    onMove?: (positon: { x: number; y: number } | undefined) => void;
    fps?: number;
  },
): { x: number; y: number } | undefined {
  const [mousePosition, setMousePosition] = useState<{ x: number; y: number } | undefined>();

  // Keep track of most recent `onMove` handler so we don't need to recreate `throttledUpdate` if it changes.
  const onMoveRef = useRef(options?.onMove);
  onMoveRef.current = options?.onMove;

  const throttledUpdate = useMemo(() => {
    const fps = options?.fps || 30;
    return _.throttle((position: { x: number; y: number } | undefined) => {
      setMousePosition(position);
      onMoveRef.current?.(position);
    }, 1000 / fps);
  }, [options?.fps]);

  useGesture(
    {
      onHover: (state) => {
        console.log(`onHover: ${state.hovering}`);
        if (!state.hovering) {
          throttledUpdate(undefined);
        }
      },
      onMove: (state) => {
        const el = 'current' in ref ? ref.current : ref;
        if (el) {
          const bounds = el.getBoundingClientRect();
          throttledUpdate({
            x: state.xy[0] - bounds.x,
            y: state.xy[1] - bounds.y + el.scrollTop,
          });
        }
      },
    },
    { target: ref },
  );

  return mousePosition;
}
```

In the failure case, we never see the `onHover: false` log emitted.  After setting a lot of logpoints and breakpoints and general futzing about, I can also say that in the failure case that [HoverEngine.leave()](https://github.com/pmndrs/use-gesture/blob/f1061c7d852bd12668eb5255489aa64e10eac6d3/packages/core/src/engines/HoverEngine.ts#L17) is never being called (which is really weird).

**Information:**
 - React Use Gesture version: 10.0.0-beta.13
 - React: 16.13.1
 - Device: MacBook Pro
 - OS: macOS 10.15.7
 - Browser: Chrome

**Comment by jwalton** [comment]
I think I've found the problem.  If I do this:

```ts
  const el = ref.current;

  useEffect(() => {
    if (el) {
      const handler = (e: Event) => {
        console.log(e.type);
      };

      el.addEventListener('pointerenter', handler);
      el.addEventListener('pointerleave', handler);

      return () => {
        el.removeEventListener('pointerenter', handler);
        el.removeEventListener('pointerleave', handler);
      };
    }
  }, [el]);
```
  
This works as you'd expect, printing `pointerenter` and `pointerleave` as the mouse enters and leaves the element.  But if I do this (which adds and removes the event listeners on every render, more closely mimics what use-gesture is doing):
  
```diff
  const el = ref.current;

  useEffect(() => {
    if (el) {
      const handler = (e: Event) => {
        console.log(e.type);
      };

      el.addEventListener('pointerenter', handler);
      el.addEventListener('pointerleave', handler);

      return () => {
        el.removeEventListener('pointerenter', handler);
        el.removeEventListener('pointerleave', handler);
      };
    }
-  }, [el]);
+  });
```

then sometimes I'll lose the `pointerleave` event.  It seems like the `pointerleave` event can happen inbetween the effect cleanup and the start of the new effect.

**Comment by dbismut** [comment]
Oh I see. And if you're not using the `target` option in `useGesture` and spread `{ ...bind() }` to the component, does that work? Also, would `useLayoutEffect` change anything (just curious)?

**Comment by dbismut** [comment]
Also, are you able to setup a minimal repro sandbox? As I'm trying locally with this code and I still get the right logs at all times.

```js
import React from 'react'
import { useHover } from '@use-gesture/react'

function App() {
  const [flag, set] = React.useState(false)
  const target = React.useRef()
  useHover(
    ({ active }) => {
      set(active) // triggers re-render
      console.log({ active })
    },
    { target }
  )

  return (
    <div ref={target} tabIndex={-1} className={styles.drag} style={style} />
  )
}
```

**Comment by jwalton** [comment]
Ok, I have a minimum reproduction which I *think* is this same bug, although it doesn't quite present exactly the same way.  I've tried this in both FF and Chrome:

https://codesandbox.io/s/use-gesture-react-use-hover-bug-dbeko?file=/src/App.tsx

If you hover over the blue box, and then move out to the white background again, it'll work.  If you hover over the blue box, then move your mouse to the green box, it won't.  On my machine, this is 100% reproducible.

![hover](https://user-images.githubusercontent.com/1771003/123293125-b7202780-d4e1-11eb-94d1-431005091bb1.gif)

**Comment by dbismut** [comment]
Ok that one is interesting indeed. I just pushed a fix on `10.0.0-beta.14` that should solve this.

This is due to a conflict between the `Move` and `Hover` gestures that both use the `pointerleave` listener: the `Move` debounces `pointermove` events to trigger `onMoveEnd` but also listens to `pointerleave` events so that it doesn't have to wait for the debounced handler to trigger `onMoveEnd`.

The problem was that `@use-gesture` was adding and removing the `Move` gesture listeners **before** the `Hover` gesture listeners. 

Here's what used to happen:

- as you guessed in an earlier message, `@use-gesture` removes and adds event listeners on every render (in theory this should make using a ref closer to using `bind()`). This allows to dynamically change the config of the gesture, but it looks like this has some side effects.
- when the mouse leaves the top div, the `Move` gesture is triggered by the `pointerleave` listener: it computes a render to update the mouse position.
- that render removes all listeners of all gestures and adds them back, but because the `Move` gesture listeners are handled **before** the `Hover` gesture listeners, the `pointerleave` listener from the `Hover` gesture was removed before it had the time to fire.
- when it's added back on the next frame it's too late, the mouse is already out of the div and there's no reason for `pointerleave` to fire again.

There must be some flaws in that logic though, as it seems that it's actually the render triggered by the bottom div that causes that handler confusion (as you've noticed, moving from the top div to the background didn't seem to trigger that bug).

Anyway, I've fixed this by handling `Hover` before any other gesture, so we should be fine for now.
